include/chainbase/shared_cow_string.hpp

```diff
@@ -21,123 +24,230 @@ namespace chainbase {
-            other._data = nullptr;
+
+      shared_cow_string(shared_cow_string&& o) noexcept {
+         if (get_allocator() == o.get_allocator()) {
+            _data = o._data;
+            o._data = nullptr;
+         } else {
+            if (o._data)
+               new (this) shared_cow_string(o.data(), o.size());
          }
-         return *this;
       }
+
       ~shared_cow_string() {
          dec_refcount();
+         _data = nullptr;
       }
-      void resize(std::size_t new_size, boost::container::default_init_t) {
-         impl* new_data = (impl*)&*_alloc.allocate(sizeof(impl) + new_size + 1);
-         new_data->reference_count = 1;
-         new_data->size = new_size;
-         new_data->data[new_size] = '\0';
-         dec_refcount();
-         _data = new_data;
+
+      shared_cow_string& operator=(const shared_cow_string& o) {
+         if (this != &o) {
+            if (get_allocator() == o.get_allocator()) {
+               dec_refcount();
+               _data = o._data;
+               if (_data != nullptr) 
+                  ++_data->reference_count;
+            } else {
+               assign(o.data(), o.size());
+            }
+         }
+         return *this;
+      }
+
+      shared_cow_string& operator=(shared_cow_string&& o)  noexcept {
+         if (this != &o) {
+            if (get_allocator() == o.get_allocator()) {
+               dec_refcount();
+               _data = o._data;
+               o._data = nullptr;
+            } else {
+               assign(o.data(), o.size());
+            }
+         }
+         return *this;
+      }
+
+      shared_cow_string&  operator=(std::string_view sv) {
+         if (!copy_in_place(sv.data(), sv.size())) {
+            _alloc(sv.data(), sv.size());
+         }
+         return *this;
       }
+
       template<typename F>
       void resize_and_fill(std::size_t new_size, F&& f) {
-         resize(new_size, boost::container::default_init);
+         if (!copy_in_place(nullptr, new_size)) {
+            _alloc(nullptr, new_size);
+         }
          static_cast<F&&>(f)(_data->data, new_size);
       }
+
       void assign(const char* ptr, std::size_t size) {
-         impl* new_data = (impl*)&*_alloc.allocate(sizeof(impl) + size + 1);
-         new_data->reference_count = 1;
-         new_data->size = size;
-         if(size)
-            std::memcpy(new_data->data, ptr, size);
-         new_data->data[size] = '\0';
-         dec_refcount();
-         _data = new_data;
+         if (!copy_in_place(ptr, size)) {
+            _alloc(ptr, size);
+         }
       }
+
       void assign(const unsigned char* ptr, std::size_t size) {
-         assign((char*)ptr, size);
+         assign((const char*)ptr, size);
       }
-      const char * data() const {
-         if (_data) return _data->data;
-         else return nullptr;
+
+      const char* data() const {
+         return _data ? _data->data : nullptr;
       }
+
+      char* data() {
```

Comment:
Maybe return a `std::span`, also consider renaming to `mutable_data()` or something other than just `data()` to make sure think twice before using.

Reply:
I renamed to `mutable_data()`. 
I did not return a `string_view` because `string_view` represent *constant* sequences of characters. I thought of using `std::span` but right now our CI for libtester compiles these headers in c++17 mode so this wouldn't work.